### PR TITLE
implements intersect node

### PR DESCRIPTION
**Pull request by yangzhg**
#2845
imlement of the intersect node
now can support  statement like `select a from t  intersect select b from t1 intersect select 1 `

### CODE REVIEW

File: be/src/exec/intersect_node.cpp [hunk 1]
```diff
@@ -56,4 +43,177 @@ Status IntersectNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+                }
+                VLOG_ROW << "hash table content: "
+                         << _hash_tbl->debug_string(true, &child(0)->row_desc());
+                build_batch.reset();
+
+                if (eos) {
+                    break;
+                }
+            }
+        } else {
```
Comment (morningman) [inline be/src/exec/intersect_node.cpp]: the `_children.size()` is 2, so is it possible to enter this `else`?

File: be/src/exec/intersect_node.cpp [hunk 2]
```diff
@@ -56,4 +43,177 @@ Status IntersectNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+            _hash_tbl.reset(temp_tbl);
+        }
+        // probe
+        _probe_batch.reset(new RowBatch(child(i)->row_desc(), state->batch_size(), mem_tracker()));
+        RETURN_IF_ERROR(child(i)->open(state));
+        while (true) {
+            RETURN_IF_CANCELLED(state);
+            bool eos = true;
+            RETURN_IF_ERROR(child(i)->get_next(state, _probe_batch.get(), &eos));
+            RETURN_IF_LIMIT_EXCEEDED(state, " Intersect , while constructing the hash table.");
```
Comment (morningman) [inline be/src/exec/intersect_node.cpp]:
```suggestion
            RETURN_IF_LIMIT_EXCEEDED(state, " Intersect , while probing the hash table.");
```

File: be/src/exec/intersect_node.h [hunk 3]
```diff
@@ -38,61 +44,36 @@ class IntersectNode : public ExecNode {
...
+    int _build_tuple_size;
+    int _build_tuple_row_size;
+    std::vector<bool> _find_nulls;
 
-    /// Index of the child that needs to be closed on the next GetNext() call. Should be set
-    /// to -1 if no child needs to be closed.
-    int _to_close_child_idx;
+    // for right outer joins, keep track of what's been joined
+    typedef boost::unordered_set<TupleRow*> TempTupleRowSet;
+    TempTupleRowSet _temp_rows;
```
Comment (morningman) [inline be/src/exec/intersect_node.h]: Not used?

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/exec/intersect_node.cpp [hunk 4]
```diff
@@ -17,36 +17,21 @@
...
     DCHECK(tnode.__isset.intersect_node);
     DCHECK_EQ(_conjunct_ctxs.size(), 0);
-    // Create const_expr_ctx_lists_ from thrift exprs.
-    auto& const_texpr_lists = tnode.intersect_node.const_expr_lists;
-    for (auto& texprs : const_texpr_lists) {
-        std::vector<ExprContext*> ctxs;
-        RETURN_IF_ERROR(Expr::create_expr_trees(_pool, texprs, &ctxs));
-        _const_expr_lists.push_back(ctxs);
-    }
+    DCHECK_GE(_children.size(), 2);
```
Comment (wutiangan) [inline be/src/exec/intersect_node.cpp]: why not supoort 3 or more？ like union？

Comment (wutiangan) [inline be/src/exec/intersect_node.cpp]: support intersect of 3 tables or more table？

File: be/src/exec/intersect_node.cpp [hunk 5]
```diff
@@ -17,36 +17,21 @@
...
+    DCHECK_EQ(tnode.intersect_node.const_expr_lists.size(), 0);
```
Comment (wutiangan) [inline be/src/exec/intersect_node.cpp]: const_expr_list not use? why not delete it?

File: be/src/exec/intersect_node.cpp [hunk 6]
```diff
@@ -56,4 +41,164 @@ Status IntersectNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+    for (size_t i = 0; i < _child_expr_lists.size(); ++i) {
+        RETURN_IF_ERROR(Expr::prepare(_child_expr_lists[i], state, child(i)->row_desc(),
+                                      expr_mem_tracker()));
+    }
+    _build_tuple_size = child(0)->row_desc().tuple_descriptors().size();
+    _build_tuple_row_size = _build_tuple_size * sizeof(Tuple*);
+    _build_tuple_idx.reserve(_build_tuple_size);
+
+    for (int i = 0; i < _build_tuple_size; ++i) {
+        TupleDescriptor* build_tuple_desc = child(1)->row_desc().tuple_descriptors()[i];
```
Comment (wutiangan) [inline be/src/exec/intersect_node.cpp]: child（1）？

File: be/src/exec/intersect_node.cpp [hunk 7]
```diff
@@ -56,4 +41,164 @@ Status IntersectNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+    // open result expr lists.
+    for (const vector<ExprContext*>& exprs : _child_expr_lists) {
+        RETURN_IF_ERROR(Expr::open(exprs, state));
+    }
+
+    for (int i = 0; i < _children.size(); ++i) {
+        // initial build hash table
+        if (_hash_tbl.get() == nullptr || _hash_tbl->size() == 0) {
+            _hash_tbl.reset(new HashTable(_child_expr_lists[0], _child_expr_lists[i],
+                                          _build_tuple_size, false, _find_nulls, id(),
```
Comment (wutiangan) [inline be/src/exec/intersect_node.cpp]: why use _build_tuple_size? if 100 table intersect, the build_tuple_size = 100?

File: be/src/exec/intersect_node.cpp [hunk 8]
```diff
@@ -56,4 +41,164 @@ Status IntersectNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+                                          mem_tracker(), 1024));
+            RowBatch build_batch(child(0)->row_desc(), state->batch_size(), mem_tracker());
```
Comment (wutiangan) [inline be/src/exec/intersect_node.cpp]: if i > 0, we can also enter this branch,  child(0) may be wrong?

File: be/src/exec/intersect_node.cpp [hunk 9]
```diff
@@ -56,4 +41,164 @@ Status IntersectNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+                break;
+            }
+        } else {
+            // if this hash table is probed rebuild it
+            std::unique_ptr<HashTable> temp_tbl(
+                    new HashTable(_child_expr_lists[0], _child_expr_lists[i], _build_tuple_size,
+                                  false, _find_nulls, id(), mem_tracker(), 1024));
+            _hash_tbl_iterator = _hash_tbl->begin();
+            while (_hash_tbl_iterator.has_next()) {
+                if (_hash_tbl_iterator.matched()) {
```
Comment (wutiangan) [inline be/src/exec/intersect_node.cpp]: when i = 0, the first table build  _hash_tbl , but you can't set matched() function.
when i = 1, the _hash_tbl's " _hash_tbl_iterator.matched() "is true of false?

File: be/src/exec/intersect_node.cpp [hunk 10]
```diff
@@ -56,4 +41,164 @@ Status IntersectNode::init(const TPlanNode& tnode, RuntimeState* state) {
...
+                    temp_tbl->insert(_hash_tbl_iterator.get_row());
+                }
+                _hash_tbl_iterator.next<false>();
```
Comment (wutiangan) [inline be/src/exec/intersect_node.cpp]: do you test the null value? can null  output.

**Review by wutiangan — COMMENTED** [review]
you'd better add some comments in key function

File: be/src/exec/intersect_node.cpp (hunk 7 above)
Comment (yangzhg) [reply be/src/exec/intersect_node.cpp]: it means number of Tuples in the build tuple row  not table count

File: be/src/exec/intersect_node.cpp (hunk 8 above)
Comment (yangzhg) [reply be/src/exec/intersect_node.cpp]: no  it always use the first table‘s row desc

**Review by morningman — APPROVED** [review]
LGTM

**Review by chaoyli — APPROVED** [review]

**Pull request merged**